File: NetworkSimulator.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import Network
import collections
import sys
import numpy as np
import numpy.random as random
import time

def logToCount(logList, countList, MAX_COUNT, STEP=1000):
  timestamps = []
  for line in logList:
    time = line[2]
    timestamps.append(time)
  
  for i in range(MAX_COUNT):
    countList.append(0)
  for t in timestamps:
    index = int(t/STEP)
    countList[index] += 1

start = time.time()

myNet = Network.Network()
myNet.loadRelationFile('./topology/dataset_100a.txt')
myNet.searchTopAS()
myNet.exchangeRoutingTable()
myNet.distributeNetAddr()

elapsed = time.time() - start
print "routing finished"
print "%f[sec]" % (elapsed)

# 各ネットワークから一定の正規確率でtracerouteを実行
for i in range(1):
  for net in myNet.as_list:
    noiseCounts = np.random.poisson(80, 300)
    for j in range(0, len(noiseCounts)):
      startTime = j * 1000
      for k in range(noiseCounts[j]):
        dstNet = random.randint(1, 30 + 1)
        dstHost = random.randint(2, 254 + 1)
        dstAddr = dstNet << 16 | dstHost
        srcAddr = net.networkAddress | 0x02
        elapsed = myNet.traceroute(srcAddr, dstAddr, startTime)

elapsed = time.time() - start
print "traceroute finished"
print "%f[sec]" % (elapsed)

File: bin/NetworkSimulator.cpp
#include "Network.cpp"
#include <time.h>
#include <cstdlib>

int main() {
  string filename = "./topology/dataset_100a.txt";
  int start = clock();
  int elapsed = 0;
  Network myNet = Network::Network();
  myNet.loadFile(filename);
  myNet.searchTopAs();
  myNet.exchangeRouteingTable();
  myNet.distributeNetAddr();
  
  elapsed = clock() - start;
  cout << "routing finished" << endl;
  cout << ((double)elapsed/(double)CLOCKS_PER_SEC) << "SEC" << endl;

  AS *node;
  map<int, AS &>::iterator  itr = myNet.asDict.begin();
  int i, j, startTime;
  unsigned int dstNet, dstHost, dstAddr, srcAddr;
  while (itr != myNet.asDict.end()) {
    node = &(itr->second);
    for (i=0; i<300; i++) {
      startTime = i * 1000;
      for (j=0; j<80; j++) {
        dstNet = rand()%100+1;
        dstHost = rand()%254+1;
        dstAddr = dstNet<<16 | dstHost;
        srcAddr = node->networkAddress;
        elapsed = myNet.traceroute(srcAddr, dstAddr, startTime, 3);
      }
    }
    itr++;
  }

  elapsed = clock() - start;
  cout << "traceroute finished" << endl;
  cout << ((double)elapsed/(double)CLOCKS_PER_SEC) << "SEC" << endl;
  return 0;
}
